components/tidb_query_expr/src/impl_string.rs

```diff
@@ -51,6 +51,42 @@ pub fn unhex(arg: BytesRef, writer: BytesWriter) -> Result<BytesGuard> {
     Ok(writer.write(hex::decode(padded_content).ok()))
 }
 
+#[inline]
+fn find_str(text: &str, pattern: &str) -> Option<usize> {
+    twoway::find_str(text, pattern).map(|i| text[..i].chars().count())
+}
+
+#[rpn_fn]
+#[inline]
+pub fn locate_2_args_utf8(substr: BytesRef, s: BytesRef) -> Result<Option<i64>> {
+    let substr = str::from_utf8(substr).unwrap();
+    let s = str::from_utf8(s).unwrap();
+    Ok(Some(
+        find_str(&s.to_lowercase(), &substr.to_lowercase()).map_or(0, |i| 1 + i as i64),
+    ))
+}
+
+#[rpn_fn]
+#[inline]
+pub fn locate_3_args_utf8(substr: BytesRef, s: BytesRef, pos: &Int) -> Result<Option<i64>> {
+    if *pos < 1 {
+        return Ok(Some(0));
+    }
+    let substr = str::from_utf8(substr).unwrap();
+    let s = str::from_utf8(s).unwrap();
+    Ok(Some(
```

Comment:
split this looog expression into two. e.g.  Store the offset of s as an intermediate variable.

Reply:
done